src/metpy/calc/thermo.py

```diff
@@ -4205,6 +4206,223 @@ def k_index(pressure, temperature, dewpoint, vertical_dim=0):
+    ...      40., 30., 25., 20.] * units.hPa
+    >>> # temperature
+    >>> T = [29.3, 28.1, 23.5, 20.9, 18.4, 15.9, 13.1, 10.1, 6.7, 3.1,
+    ...      -0.5, -4.5, -9.0, -14.8, -21.5, -29.7, -40.0, -52.4,
+    ...      -59.2, -66.5, -74.1, -78.5, -76.0, -71.6, -66.7, -61.3,
+    ...      -56.3, -51.7, -50.7, -47.5] * units.degC
+    >>> # relative humidity
+    >>> rh = [.85, .65, .36, .39, .82, .72, .75, .86, .65, .22, .52,
+    ...       .66, .64, .20, .05, .75, .76, .45, .25, .48, .76, .88,
+    ...       .56, .88, .39, .67, .15, .04, .94, .35] * units.dimensionless
+    >>> # calculate dewpoint
+    >>> Td = dewpoint_from_relative_humidity(T, rh)
+    >>> galvez_davison_index(p, T, Td)
+    <Quantity(-8.06886508, 'dimensionless')>
+    """
+    # Calculate mixing ratio from dewpoint in two steps
+    relative_humidity = relative_humidity_from_dewpoint(
+        temperature, dewpoint
+    )
+    mixing_ratio = mixing_ratio_from_relative_humidity(
+        pressure, temperature, relative_humidity
+    )
+
+    # Calculate potential temperature
+    potential_temp = potential_temperature(pressure, temperature)
+
+    if np.any(np.max(pressure, axis=vertical_dim) < 950 * units.hectopascal):
+        indices_without_950 = np.where(
+            np.max(pressure, axis=vertical_dim) < 950 * units.hectopascal
+        )
+        raise ValueError(
+            f'Data not provided for 950hPa or higher pressure. '
+            f'GDI requires 950hPa temperature and dewpoint data, '
+            f'see referenced paper section 3.d. in docstring for discussion of'
+            f' extrapolating sounding data below terrain surface in high-'
+            f'elevation regions.\nIndices without a 950hPa or higher datapoint'
+            f':\n{indices_without_950}'
+            f'\nMax provided pressures:'
+            f'\n{np.max(pressure, axis=0)[indices_without_950]}'
+        )
+    # Convert temperature before interpolation
+    temperature_k = temperature.to('kelvin')
+    potential_temp_k = potential_temp.to('kelvin')
+
+    # Interpolate to find temperature and mixing ratio at 950, 850, 700, and 500 hPa
+    (
+        (t950, t850, t700, t500),
+        (r950, r850, r700, r500),
+        (th950, th850, th700, th500)
+    ) = interpolate_1d(
+        units.Quantity([950, 850, 700, 500], 'hPa'),
+        pressure, temperature_k, mixing_ratio, potential_temp_k,
+        axis=vertical_dim,
+    )
+
+    th_a = th950
+    r_a = r950
+
+    th_b = 0.5 * (th850 + th700)
+    r_b = 0.5 * (r850 + r700)
+
+    th_c = th500
+    r_c = r500
+
+    alpha = -10 * units.kelvin  # Empirical adjustment
+
+    # Latent heat of vaporization of water - different from MetPy constant
+    #  Using different value, from paper, since GDI unlikely to be used to
+    #  derive other metrics
+    l_0 = 2.69E6 * (units.joule / units.kilogram)
+
+    # Temperature math from here on requires kelvin units
+    eptp_a = th_a * np.exp((l_0 * r_a) / (Cp_d * t850))
```

Comment:
Note: The Cp_d value from the paper is 1005.7 J/kg/K, which is different from the value in MetPy which is 1004.666... J/kg/K. Not a huge difference and with their calculation of potential temperature with the 2/7 ratio I think it all falls out in the wash. I am for keeping the use of MetPy Cp_d value and using our potential temperature calculation in this function.